### ISSUE TITLE
Include / Import External Files (split Job in more files)

### ISSUE THREAD

**Issue by pcbl** [hashicorp/nomad#7954]
Hi Folks,

I must start with a disclaimer, mentioning that I am new to NOMAD and might be missing something.

The question is simple:
Does HCL support some sort of include/import, so that I could split a job file among several files, for the sake of maintenance?

The scenario:
I do have one specific nomad job wich includes several Tasks. Those have some level of dependency among them, therefore we keep it within a single job. As the file is becoming somehow too big, we would like to split every Task within a separate file, including all of them on the main Job file.

Would that be possible? I have searched a little bit and found this issue, which seems somehow similar, but it is quite old (2017).
https://github.com/hashicorp/hcl/issues/187

Maybe there´s a good way to do it...

Would anyone has a suggestion?

BR
Pedro

**Comment by tgross** [comment]
Hi @pcbl! This isn't something you can do today with Nomad's job files. We do support it for [Nomad configuration files](https://www.nomadproject.io/docs/configuration/#load-order-and-merging). We have plans to implement [HCL2](https://github.com/hashicorp/nomad/issues/7220) in the near future and I think this would be an interesting feature to look into.

**Comment by pcbl** [comment]
Ok @tgross , thanks for the heads up. I will keep an eye here case this feature comes up. By now I will end up build some solution to merge the files myself before I start the job. Thx!

**Comment by radeksimko** [comment]
As part of looking into this I'd also recommend standardizing `nomad` as the file extension for job specs, or coming up with an extension which can even differentiate agent config from a job spec - e.g. `*.job.nomad` vs `*.agent.nomad`.

Product-specific extension (unlike generic `.hcl`) enables editors to provide more accurate highlighting, completion etc. and don't have to guess what flavour of HCL it is based on content and what schema to expect.

**Comment by sofixa** [comment]
So now with HCL2 it's possible to include files: 

https://www.nomadproject.io/docs/job-specification/hcl2/functions/file/file

But without any templating/rendering, unlike terraform which has [templatefile](https://www.terraform.io/docs/configuration/functions/templatefile.html), which can render a template from a file. Are there plans to add this to Nomad's HCL2 implementation ?

**Comment by Oloremo** [comment]
Looking for that as well and actually not sure how to use https://www.nomadproject.io/docs/job-specification/hcl2/functions/file/file to split the HCL2.

Can I do something like:
```
job "name" {
  group "name" {
    task "name1" {
      file("path/to/file/with/all/stanzas/from/that/task")
    }
    task "name2" {
      file("path/to/file/with/all/stanzas/from/that/task")
    }
  }
}
```
?

**Comment by str-dst** [comment]
This would be really useful since large job templates quickly become confusing and hard to maintain.

And the 64kib size limit for vars on the API makes it impossible to create templates beyond a certain size.

**Comment by next-jesusmanuelnavarro** [comment]
> This would be really useful since large job templates quickly become confusing and hard to maintain.

Yes they are.  In fact, I'd say it leads to the *"antipattern"* of deploying as multiple jobs what naturally should be one.  i.e.: a typical three-tier product with front/middle/back containers that would naturally fit into a single job with multiple task groups, ends up being three jobs in order for the hcl to be manageable by humans... at the expense of its deployment orchestration having to be delegated somewhere else.

This can be *workarounded* by using a third-party templating mechanism. I.e.: we use Ansible's jinja2 to render hcl from hcl.j2 templates.  This approach gives us the ability to use environment-tied variables *"for free"* by means of Ansible inventories, and can be easily expanded to use multi-file templates (aka includes).

> And the 64kib size limit for vars on the API makes it impossible to create templates beyond a certain size.

...Yes, up to the point you reach this limit (it recently was my case while working on a PoC when we needed to overwrite container entrypoints for faster testing)

So, a *"full"* solution for this issue would be not only adding the ability ot include multiple files into one but also some feature to allow for the API server to cope with larger jobs (i.e.: multipart uploads / server-side ability to *"build"* a single jobs out of many separated calls.
